integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/createTable/TestNonTransactionalCarbonTableWithAvroDataType.scala

```diff
@@ -884,4 +962,310 @@ class TestNonTransactionalCarbonTableWithAvroDataType extends QueryTest with Bef
+        |	"fields": [
+        |		{
+        |			"name": "dec_field",
+        |						"type": {"type" : "bytes",
+        |                     "logicalType": "decimal",
+        |                     "precision": 5,
+        |                     "scale": 2
+        |                    }
+        |}
+        |	]
+        |}""".stripMargin
+    val nn = new org.apache.avro.Schema.Parser().parse(schema1)
+    val logicalType = LogicalTypes.decimal(5, 2)
+    val decimal = new java.math.BigDecimal("1218").setScale(2)
+    //get unscaled 2's complement bytearray
+    val bytes =
+      decimalConversion.toBytes(decimal, nn.getField("dec_field").schema, logicalType)
+    val data = DatatypeConverter.printBase64Binary(bytes.array())
+    val json1 =
+      s"""{"dec_field":"$data"}""".stripMargin
+    val avroRec = new GenericData. Record(nn)
+    avroRec.put("dec_field", bytes)
+    val exception1 = intercept[Exception] {
+    val writer = CarbonWriter.builder
+      .outputPath(writerPath).isTransactionalTable(false).buildWriterForAvroInput(nn)
+    writer.write(avroRec)
+    writer.close()
+    }
+    assert(exception1.getMessage
+      .contains("Data Loading failed as value Precision 6 is greater than specified Precision 5 in Avro Schema"))
+  }
+
+  test("test union with multiple record type") {
+    sql("drop table if exists sdkOutputTable")
+    FileFactory.deleteAllCarbonFilesOfDir(FileFactory.getCarbonFile(writerPath))
+    val schema1 =
+      """{
+        |	"namespace": "test.avro",
+        |	"type": "record",
+        |	"name": "NewCar2",
+        |	"fields": [
+        |      {
+        |		  "name": "optionalExtra",
+        |    "type": ["null",{
+        |       "type":"record",
+        |       "name":"Stereo",
+        |       "fields" :[{
+        |       "name":"make",
+        |       "type":"string"
+        |       },
+        |       {
+        |       "name":"speakers",
+        |       "type":"int"
+        |       }]
+        |       },{
+        |       "type":"record",
+        |       "name":"LeatherTrim",
+        |       "fields":[{
+        |       "name":"colour",
+        |       "type":"string"
+        |       }]
+        |       }],
+        |       "default":null
+        |       }]
+        |
+        |}""".stripMargin
+
+    val json1 =
+      """{"optionalExtra":{"test.avro.LeatherTrim":{"colour":"ab"}}}""".stripMargin
+
+    val nn = new org.apache.avro.Schema.Parser().parse(schema1)
+    val record = testUtil.jsonToAvro(json1, schema1)
+
+    val writer = CarbonWriter.builder
+      .outputPath(writerPath).isTransactionalTable(false).buildWriterForAvroInput(nn)
+    writer.write(record)
+    writer.close()
+    sql(
+      s"""CREATE EXTERNAL TABLE sdkOutputTable STORED BY
+         |'carbondata' LOCATION
+         |'$writerPath' """.stripMargin)
+    checkAnswer(sql("select * from sdkOutputTable"),
+      Seq(Row(Row(Row(null,null),Row("ab")))))
+  }
+
+  test("test union with multiple Enum type") {
```

Comment:
Can you add test case to read using data source file format. (Syntax "using carbon" with schema Refer SparkCarbonDataSourceTest ).
This will help users on how to define schema for avro logcial types & union .